src/open_dread_rando/pickups/split_pickups.py

```diff
@@ -0,0 +1,394 @@
+    def as_list(self) -> list[dict]:
+        ret = [self.add_primary_gun.raw]
+        if self.set_billboard_group_params is not None:
+            ret.append(self.set_billboard_group_params.raw)
+        if self.set_gun_charge_params is not None:
+            ret.append(self.set_gun_charge_params.raw)
+        return ret
+
+    @property
+    def name(self) -> str:
+        return self.add_primary_gun.params.name
+
+    @name.setter
+    def name(self, value: str):
+        self.add_primary_gun.params = dataclasses.replace(
+            self.add_primary_gun.params,
+            name=value,
+        )
+        if self.set_billboard_group_params is not None:
+            self.set_billboard_group_params.name = value
+        if self.set_gun_charge_params is not None:
+            self.set_gun_charge_params.name = value
+
+
+
+class AddSecondaryGun(CCFunc):
+    pass
+
+
+SAMUS_BMSAD_PATH = "actors/characters/samus/charclasses/samus.bmsad"
+
+def patch_split_pickups(editor: PatcherEditor):
+    samus = editor.get_file(SAMUS_BMSAD_PATH, Bmsad)
+
+    primaries = _patch_split_beams(editor)
+    secondaries = _patch_split_missiles(editor)
+    selections = [
+        {
+            "name": "SelectGun",
+            "unk": 1,
+            "params": {
+                "Param1": {
+                    "type": "s",
+                    "value": "PowerBeam",
+                },
+                "Param2": {
+                    "type": "b",
+                    "value": "True"
+                }
+            }
+        },
+        {
+            "name": "SelectGun",
+            "unk": 1,
+            "params": {
+                "Param1": {
+                    "type": "s",
+                    "value": "Missile",
+                },
+                "Param2": {
+                    "type": "b",
+                    "value": "False"
+                }
+            }
+        }
+    ]
+
+    samus.raw.property.components["GUN"].functions = primaries + secondaries + selections
+
+
+def _patch_split_beams(editor: PatcherEditor) -> list[dict]:
+    samus = editor.get_file(SAMUS_BMSAD_PATH, Bmsad)
+    gun_funcs = samus.raw.property.components["GUN"].functions
+
+    power = PrimaryGunFuncs(
+        gun_funcs[0],
+        gun_funcs[1],
+        gun_funcs[2],
+    )
+    wide = PrimaryGunFuncs(
+        gun_funcs[3],
+        gun_funcs[4],
+        gun_funcs[5],
+    )
+    plasma = PrimaryGunFuncs(
+        gun_funcs[6],
+        gun_funcs[7],
+        gun_funcs[8],
+    )
+    wave = PrimaryGunFuncs(
+        gun_funcs[9],
+        gun_funcs[10],
+        gun_funcs[11],
+    )
+    hyper = PrimaryGunFuncs(
+        gun_funcs[12],
+        gun_funcs[13],
+        None,
+    )
+    grapple = PrimaryGunFuncs(
+        gun_funcs[14],
+        None,
+        None,
+    )
+    spbgun = PrimaryGunFuncs(
+        gun_funcs[15],
+        None,
+        None,
+    )
+
+    power.add_primary_gun.params = dataclasses.replace(
+        power.add_primary_gun.params,
```

Comment:
this returns the container, not the dataclass. I think it's being overriden by raw.params or something.

Reply:
hacky fix to resolve this: add this to PrimaryGunFuncs

```def __post_init__(self):
        self.add_primary_gun = AddPrimaryGun(self.add_primary_gun)
        if self.set_billboard_group_params:
            self.set_billboard_group_params = SetBillBoardGroupParams(self.set_billboard_group_params)
        if self.set_gun_charge_params:
            self.set_gun_charge_params = SetGunChargeParams(self.set_gun_charge_params)```